### DISCUSSION TITLE
Implementation Question RE: Star Force

### DISCUSSION THREAD

**Discussion by AidyCC** [mamedev/discussions#10] [Q&A]
Hi

I posted this question in the Issues area by mistake was pointed to the source code [here](https://github.com/mamedev/mame/blob/master/src/devices/machine/z80ctc.cpp).  I was pointed to this forum so I thought I’d ask the question here for any other insight…

I'd like to get a better understanding of how the sound works in Star Force (driver: senjyo.cpp).

The hardware has 3 SN76496 chips, each of which has 3 channels giving the hardware 9 channels in total.

Analyzing the code from the SF sound ROM shows that there are actually 10 sound channels used in the game. The first 9 are the ones provided by the sound chips but the last one gets it's frequency from the z80ctc chip on port 0x0a.

My question is where in the emulation code is that 10th channel managed and the sound generated?  I can’t immediately see where that is from the z80ctc source code linked above.  There appears to be no link between the z80ctc code and the sound output.

Thanks.

**Comment by Osso13** [comment]
As far as I understand it, the Z80CTC here: https://github.com/mamedev/mame/blob/master/src/mame/tecmo/senjyo.cpp#L573 drives a DAC implemented here https://github.com/mamedev/mame/blob/master/src/mame/tecmo/senjyo_a.cpp#L22

**Comment by happppp** [comment]
Those links are outdated, sorry ;)
The DAC sound emulation is now in the senjyo.cpp file.
I improved the emulation a bit. The CTC clocks some sort of address counter, which connects to a small PROM, which writes to a DAC. It's a saw wave(kind of).

**Comment by cuavas** [comment]
It’s what would be called a “digital oscillator” if it were in a synth (for example the DW-8000/EX-8000 use this approach for generating the base waveforms).  It’s a straightforward way to create an arbitrary repeating waveform at variable frequency.

**Comment by AidyCC** [comment]
Wow!  Thanks for the information, it's really helped...

A related question, one of the Sound Effects played through the DAC is a siren sound where the values 0x02 .. 0x1e are sent to the z80ctc and loop repeatedly.

Is there a calculation to determine the actual frequency of the tones played?

Thanks,

-Adrian

**Comment by happppp** [comment]
The Z80 CTC is a programmable timer. So yeah, the game sets up the timer and the output frequency changes.
Something like: 2mhz / prescaler / timerval / 8?
prescaler is probably 256.
